lib/rubocop/cop/rspec/stubbed_mock.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks that message expectations do not have a configured response.
+      #
+      # @example
+      #
+      #   # bad
+      #   expect(foo).to receive(:bar).with(42).and_return("hello world")
+      #
+      #   # good (without spies)
+      #   allow(foo).to receive(:bar).with(42).and_return("hello world")
+      #   expect(foo).to receive(:bar).with(42)
+      #
+      class StubbedMock < Base
+        MSG = 'Prefer `%<replacement>s` to `%<method_name>s` when ' \
+              'configuring a response.'
+
+        def_node_matcher :message_expectation?, <<-PATTERN
+          {
+            (send nil? { :receive :receive_message_chain } ...)
+            (send (send nil? :receive ...) :with ...)
+          }
+        PATTERN
+
+        def_node_matcher :configured_response?, <<~PATTERN
+          { :and_return :and_raise :and_throw :and_yield
+            :and_call_original :and_wrap_original }
+        PATTERN
+
+        def_node_matcher :expectation, <<~PATTERN
+          (send
+            $(send nil? $#{Expectations::ALL.node_pattern_union} ...)
+            :to $_)
+        PATTERN
+
+        def_node_matcher :matcher_with_configured_response, <<~PATTERN
+          (send #message_expectation? #configured_response? _)
+        PATTERN
+
+        def_node_matcher :matcher_with_return_block, <<~PATTERN
+          (block #message_expectation? args _)
+        PATTERN
+
+        def_node_matcher :matcher_with_blockpass_or_hash, <<~PATTERN
+          {
+            (send nil? { :receive :receive_message_chain } ... block_pass)
+            (send (send nil? :receive ...) :with ... block_pass)
+            (send nil? :receive_messages hash)
+            (send nil? :receive_message_chain ... hash)
+          }
+        PATTERN
+
+        def on_send(node)
+          expectation(node, &method(:on_expectation))
+        end
+
+        private
+
+        def on_expectation(expectation, method_name, matcher)
```

Comment:
Maybe I’m having a bad day, but I have a _really_ hard time understanding this code. It’s not immediately obvious what the node matchers actually match. The names are easy enough to understand (maybe except `matcher_with_blockpass_or_hash` which I’m not sure I get), but… Maybe we should start adding adding comments to our node matchers, with one or two examples of what the would match?

It can also be that I lack training in reading node patterns (and `parser` output). I wonder how others feel about this?

Reply:
Thanks for this fresh view on the cop.
It's quite common to miss the complexity when working on it, but as time goes, this complexity becomes apparent.

In addition to node matcher YARD docs that I believe now help a lot to understand [`RSpec/EmptyExampleGroup`'s node patterns](https://github.com/rubocop-hq/rubocop-rspec/blob/f1048eb8389d44d47afeff60e6db797487bbc46a/lib/rubocop/cop/rspec/empty_example_group.rb#L68), there's also [Node pattern comment feature](https://docs.rubocop.org/rubocop-ast/node_pattern.html#comments).
I intend to add both to this cop's node patterns.